noisy-simulator/src/density_matrix_simulator.rs

```diff
@@ -0,0 +1,346 @@
+    dimension: usize,
+    /// Number of qubits in the system.
+    number_of_qubits: usize,
+    /// Theoretical change in trace due to operations that have been applied so far.
+    trace_change: f64,
+    /// Vector storing the entries of the density matrix.
+    data: ComplexVector,
+}
+
+impl DensityMatrix {
+    fn new(number_of_qubits: usize) -> Self {
+        let dimension = 1 << number_of_qubits;
+        let mut data = ComplexVector::zeros(dimension * dimension);
+        data[0].re = 1.0;
+        Self {
+            dimension,
+            number_of_qubits,
+            trace_change: 1.0,
+            data,
+        }
+    }
+
+    /// Builds a `DensityMatrix` from its raw fields. Returns `None` if
+    ///  the provided args don't represent a valid `DensityMatrix`.
+    ///
+    /// This method is to be used from the PyO3 wrapper.
+    pub fn try_from(
+        dimension: usize,
+        number_of_qubits: usize,
+        trace_change: f64,
+        data: ComplexVector,
+    ) -> Option<Self> {
+        if 1 << number_of_qubits != dimension || data.len() != dimension * dimension {
+            None
+        } else {
+            Some(Self {
+                dimension,
+                number_of_qubits,
+                trace_change,
+                data,
+            })
+        }
+    }
+
+    /// Returns a reference to the vector containing the density matrix's data.
+    pub fn data(&self) -> &ComplexVector {
+        &self.data
+    }
+
+    /// Returns dimension of the matrix. E.g.: If the matrix is 5 x 5, then dimension is 5.
+    pub fn dimension(&self) -> usize {
+        self.dimension
+    }
+
+    /// Returns the number of qubits in the system.
+    pub fn number_of_qubits(&self) -> usize {
+        self.number_of_qubits
+    }
+
+    /// Returns `true` if the matrix is Hermitian.
+    fn is_hermitian(&self) -> bool {
+        for row in 0..self.dimension {
+            for col in 0..self.dimension {
+                let elt = self.data[self.dimension * row + col];
+                let mirror_elt = self.data[self.dimension * col + row];
+                if (elt.re - mirror_elt.re).abs() > TOLERANCE
+                    || (elt.im + mirror_elt.im).abs() > TOLERANCE
+                {
+                    return false;
+                }
+            }
+        }
+        true
+    }
+
+    /// Returns `true` if the trace of the matrix is 1.
+    fn is_normalized(&self) -> bool {
+        (self.trace() - 1.0).abs() <= TOLERANCE
+    }
+
+    /// Returns the trace of the matrix. The trace is the sum of the diagonal entries of a matrix.
+    fn trace(&self) -> f64 {
+        let mut trace: Complex<f64> = Complex::ZERO;
+        for idx in 0..self.dimension {
+            trace += self.data[(self.dimension + 1) * idx];
+        }
+        assert!(
+            trace.im <= TOLERANCE,
+            "state trace is not real, imaginary part is {}",
+            trace.im
+        );
```

Comment:
Why would we crash the user in this case?

Reply:
Yes you are right, fixed this